### PR TITLE
Give Ollama responses room in the context window

### PR DESCRIPTION
**Pull request by franzthiemann**
## Problem

Ollama's `num_ctx` bounds the **prompt and the generated answer together**, but `OllamaHandler` sized it from the prompt alone:

```
targetLength = ceil(max(estimatedTokens, default) * 1.2)
```

The `1.2` is tokenizer slack, not output room. A 10k-character prompt produced a ~4.8k window with only a few hundred tokens left to answer in, so responses were cut short well before the model's real limit.

A second bug compounded it. The ceiling was:

```ts
limit: modelInfo.contextLength || defaultContextLength
```

Whenever `ollama.show()` failed, or the returned `model_info` carried no `*.context_length` key, `contextLength` stayed `0` and the limit collapsed to **2048** — silently capping every request regardless of what the model actually supports.

## Changes

**Sizing** (`optimizeContext`) — the window is now `max(estTokens, default) * scale + OUTPUT_RESERVE_TOKENS`, clamped to the model's real context length and never shrinking below a window the model already ran with. When the real length is unknown the fallback ceiling is `FALLBACK_MAX_CONTEXT_LENGTH` (8192) rather than 2048.

Embeddings pass an output reserve of `0` — they return a vector, not tokens — and remain capped by the embedding model's own limit, so they cannot overshoot.

The computed window is always sent, including when it lands at or below `DEFAULT_CONTEXT_LENGTH`. Omitting `num_ctx` is not a request for our default — it hands the choice to Ollama's own configured default, which may be smaller and would silently shrink the window.

**Caller override respected** — `resolveContextWindow()` auto-sizes only when `options.num_ctx` is absent. An explicitly pinned window is never widened or shrunk, in `execute` and `toolsExecute` alike.

**Configurable scale** — exposed as a slider (1–4, step 0.5, default 2) under *Developer settings*, persisted as `ollamaContextScale` and clamped in the handler so a malformed value can't produce an absurd `num_ctx`. Shared constants live in a new `src/constants/ollamaContext.ts` rather than in the handler, because `settings.test.ts` mocks the handler module and importing from it breaks that suite.

**Generation-limit reporting** — when the final stream chunk carries `done_reason === 'length'`, the accumulated text is kept untouched and the incomplete answer is reported via an unconditional `console.warn` plus a `Notice`. The warning deliberately bypasses `logger`, which is gated on `debugLogging` and would have hidden it by default.

The message is neutral about the cause. Ollama returns `length` both for a filled context window and for a reached `num_predict`, and does not distinguish them — a 10-token prompt under a 32-token cap triggers it with `num_ctx = 8192`, where raising the context scale changes nothing.

There is deliberately **no prompt-truncation detection**. Ollama removes whole old messages before evaluation, so `prompt_eval_count` can sit far below `num_ctx` even when history was lost, and a prompt that exactly fills the window was not necessarily truncated. Token counts cannot support that claim in either direction, so nothing is asserted about dropped input.

## Notes for review

- **No service API change.** `ollamaContextScale` is an optional plugin setting, not part of the consumer-facing service API, so `AI_PROVIDERS_SERVICE_VERSION` and `REQUIRED_AI_PROVIDERS_VERSION` are unchanged.
- **Embedding windows do grow** where the model allows it (scale applies, reserve does not). Still bounded by the model's real context length; flagging it as a deliberate call rather than an oversight.
- New i18n key `errors.ollamaGenerationLimitReached` added to all 11 locales.
- `AGENTS.md` updated per its documentation protocol, including a standing note not to infer prompt truncation from token counts.

## Testing

`npm run lint`, `npm run typecheck`, `npm run format:check` and `npm run build` all pass. `OllamaHandler.ts`, `settings.ts` and `ollamaContext.ts` are each at 100% coverage.

Tests cover the sizing formula (output reserve, the unknown-length fallback, the no-shrink rule, scale clamping); the `num_ctx` actually passed to `ollama.chat` and `ollama.embed` for model limits 512, 1024 and 2048 across `execute`, `toolsExecute` and `embed`; an explicit `num_ctx` of 32768 and of 2048 surviving both `execute` and `toolsExecute`, verified to fail without the guard; the neutral warning on `done_reason === 'length'` including the small-`num_predict` case, with the partial text intact; and silence both when evaluation ran below the window and on an exact fit.

One caveat on the test run: `src/i18n/index.test.ts` fails on this branch, but it fails identically on a clean `master` — jsdom in this setup provides no `window.localStorage`, so that suite's `beforeEach` throws. Untouched here; the warning tests stub `localStorage` locally for the same reason.

### CODE REVIEW

File: src/handlers/OllamaHandler.ts [hunk 1]
```diff
@@ -587,18 +685,19 @@ export class OllamaHandler implements IAIHandler {
...
 
+        let num_ctx: number | undefined;
         if (processedImages.length === 0) {
             const inputLength = chatMessages.reduce(
                 (acc, msg) => acc + msg.content.length,
                 0
             );
 
-            const num_ctx = this.applyContextOptimization({
+            num_ctx = this.applyContextOptimization({
```
Comment (pfrankov) [inline src/handlers/OllamaHandler.ts]: **[P1] Preserve an explicit `options.num_ctx`.**

With a fresh handler and a 10-character prompt, the new default formula returns `6144`. This overwrites both `options.num_ctx = 32768` and `2048`; the base version preserves both for this request. This can shrink the caller's context or exceed its memory budget.

Only auto-size when `options.num_ctx` is absent, in both `execute` and `toolsExecute`. Pass the effective value sent to Ollama to the warning logic. Add regression tests for both explicit values in both paths.

File: src/handlers/OllamaHandler.ts [hunk 2]
```diff
@@ -167,6 +191,72 @@ export class OllamaHandler implements IAIHandler {
...
+        modelName: string;
+        num_ctx?: number;
+    }): void {
+        const promptTokens = chunk.prompt_eval_count;
+        const promptTruncated =
+            typeof num_ctx === 'number' &&
+            typeof promptTokens === 'number' &&
+            promptTokens >= num_ctx;
+        const outputTruncated = chunk.done_reason === 'length';
+
```
Comment (pfrankov) [inline src/handlers/OllamaHandler.ts]: **[P2] The final token count does not prove prompt truncation.**

[Ollama removes whole old messages before evaluation](https://github.com/ollama/ollama/blob/main/server/prompt.go). The remaining prompt can be well below `num_ctx`, so this check misses lost history. An exact fit also does not prove that input was removed.

Remove this definite “input was dropped” warning and the related detection claims in the documentation/PR, or replace it with a reliable check. Cover both removed history with a below-limit count and an untruncated exact-fit prompt. Changing the threshold alone cannot fix this.

File: src/i18n/en.json [hunk 3]
```diff
@@ -66,6 +68,8 @@
         "pluginMustBeUpdated": "Plugin must be updated",
         "pluginMustBeUpdatedFormatted": "Plugin must be updated to work with this version of AI Providers",
         "providerNameRequired": "Provider name is required",
-        "providerNameExists": "A provider with the name \"{{name}}\" already exists"
+        "providerNameExists": "A provider with the name \"{{name}}\" already exists",
+        "ollamaPromptTruncated": "Ollama truncated the prompt for \"{{model}}\": it did not fit in the {{contextLength}} token context window, so part of the input was dropped. Increase the Ollama context size scale in settings.",
+        "ollamaOutputTruncated": "Ollama stopped generating for \"{{model}}\": the response ran out of context. The partial answer was kept. Increase the Ollama context size scale in settings."
```
Comment (pfrankov) [inline src/i18n/en.json]: **[P2] `length` does not necessarily mean the context ran out.**

[Ollama also returns `length` when `num_predict` is reached](https://github.com/ollama/ollama/blob/main/runner/llamarunner/runner.go). A 10-token prompt and a 32-token output limit can trigger this warning with `num_ctx = 8192`. Increasing the context scale does not fix that.

Use a neutral “generation limit reached; partial answer kept” message, update all locales, and test this case. Also correct the comments/docs/PR claim that a full context always stops generation: Ollama can shift context and continue.

**Review by pfrankov — CHANGES_REQUESTED** [review]
Please fix the three issues below and add regression tests before merge.

**Remove AI authorship credits.** Amend `6b37564` to remove the `Co-Authored-By: Claude ...` trailer, then push with `--force-with-lease`. A new cleanup commit will not remove that trailer from history. Remove the “Generated with Claude Code” footer from the PR description and any equivalent AI-authorship credits in commit messages, code comments, and discussion comments you control. Keep normal references to AI providers and models.

For this project, AI is a development tool, not a co-author—just like VS Code or macOS. Human contributors remain responsible for the code, tests, and maintenance. Using a tool does not make it an author.

Reviewed `6b37564` against `9ec77ec`: code review and isolated JavaScript checks; the full project test suite was not run.

File: src/handlers/OllamaHandler.ts (hunk 1 above)
Comment (franzthiemann) [reply src/handlers/OllamaHandler.ts]: Fixed in b57f8f3. `resolveContextWindow()` returns early when `num_ctx` is already present in the request options, so auto-sizing runs only when the caller left it absent — in `execute` and `toolsExecute` alike.

Regression tests cover 32768 and 2048 across both paths. I verified they fail without the guard: removing the early return breaks all four, since a short prompt computes 6144 and overwrites the pinned value.

On passing the effective value to the warning logic: that became moot once the prompt-truncation check was removed (P2), as the remaining warning no longer reads `num_ctx`. I dropped the parameter from `streamOllamaResponse` rather than thread a value nothing consumes.

File: src/handlers/OllamaHandler.ts (hunk 2 above)
Comment (franzthiemann) [reply src/handlers/OllamaHandler.ts]: Agreed — the check was unsound, and removed entirely in b57f8f3 rather than rethresholded. You are right that `prompt.go` trims whole messages before evaluation, so `prompt_eval_count` carries no information about what was dropped, and an exact fit proves nothing either.

Gone with it: the `errors.ollamaPromptTruncated` string in all 11 locales, the `prompt_eval_count` field on the chunk type, and the detection claims in `AGENTS.md` and the PR description. I did not look for a substitute signal — there does not appear to be a reliable client-side one.

Two tests pin the silence: a below-limit count standing in for removed history, and an exact-fit prompt. `AGENTS.md` now carries a standing note not to infer prompt truncation from token counts, so this does not get reintroduced later.

File: src/i18n/en.json (hunk 3 above)
Comment (franzthiemann) [reply src/i18n/en.json]: Fixed in b57f8f3. The key is now `errors.ollamaGenerationLimitReached`, neutral about the cause:

> Ollama stopped generating for "{{model}}" after reaching a generation limit. The partial answer was kept.

No mention of the context window, and no advice to raise the scale — updated across all 11 locales. A test covers your case (10-token prompt, 32-token `eval_count`, roomy window) and asserts the message does not blame the context.

The claim that a filled window always stops generation is corrected too: the code comment on the warning now records that Ollama reports the same reason for a filled window and for `num_predict`, and `AGENTS.md` notes that a full context does not necessarily stop generation since Ollama can shift context and continue. The PR description carries the same correction.

**Comment by franzthiemann** [comment]
All four points addressed — `b57f8f3` on top of an amended `c3f9a64` (was `6b37564`), force-pushed with `--force-with-lease`.

**AI authorship credits.** The `Co-Authored-By` trailer is gone from history via amend, not a cleanup commit

**[P1]** `resolveContextWindow()` auto-sizes only when `options.num_ctx` is absent, in both paths. Regression tests for 32768 and 2048, confirmed to fail without the guard.

**[P2 prompt truncation]** Removed rather than rethresholded, along with its locale strings and the `prompt_eval_count` field. Two tests pin the silence.

**[P2 `length`]** Now a neutral "generation limit reached; partial answer kept" across all 11 locales, with the small-`num_predict` case tested. The "full context always stops generation" claim is corrected in the comments, `AGENTS.md` and the PR body.

Per-thread detail is in the replies above.

Verification: `lint`, `typecheck`, `format:check` and `build` pass; 441 tests pass with `OllamaHandler.ts`, `settings.ts` and `ollamaContext.ts` each at 100% coverage. One caveat worth repeating from the PR body — `src/i18n/index.test.ts` fails here, but identically on a clean `master` (jsdom provides no `window.localStorage`, so its `beforeEach` throws). Untouched, and happy to fix it separately if useful.

File: src/handlers/OllamaHandler.ts (hunk 1 above)
Comment (franzthiemann) [reply src/handlers/OllamaHandler.ts]: resolveContextWindow() auto-sizes only when options.num_ctx is absent, in both paths. Regression tests for 32768 and 2048, confirmed to fail without the guard.

File: src/handlers/OllamaHandler.ts (hunk 2 above)
Comment (franzthiemann) [reply src/handlers/OllamaHandler.ts]: Removed rather than rethresholded, along with its locale strings and the prompt_eval_count field. Two tests pin the silence.

File: src/i18n/en.json (hunk 3 above)
Comment (franzthiemann) [reply src/i18n/en.json]: Now a neutral "generation limit reached; partial answer kept" across all 11 locales, with the small-num_predict case tested. The "full context always stops generation" claim is corrected in the comments, AGENTS.md and the PR body.

**Review by pfrankov — DISMISSED** [review]
Re-reviewed at b57f8f3. The requested changes are addressed: explicit num_ctx values are preserved in both execute paths, unsupported prompt-truncation detection and claims are removed, done_reason=length uses neutral generation-limit messaging across all locales, and the reported cases have regression coverage. The amended history, PR description, and current changes no longer contain AI authorship credits. Ready to merge.

File: src/handlers/OllamaHandler.ts [hunk 4]
```diff
@@ -191,45 +271,52 @@ export class OllamaHandler implements IAIHandler {
...
             limit
         );
 
-        // Update only if we need context larger than previous
-        const shouldUpdate = targetLength > lastContextLength;
         return {
-            num_ctx: targetLength,
-            shouldUpdate,
+            num_ctx:
+                targetLength > defaultContextLength ? targetLength : undefined,
```
Comment (pfrankov) [inline src/handlers/OllamaHandler.ts]: **[P2] Send the computed context window even when it is 2048 or smaller.**

With no caller override, a 10,000-character prompt, `lastContextLength: 2048`, and `limit: 2048`, base `9ec77ec` returns `num_ctx: 2048`, but this code returns `undefined`. The outgoing request then omits `num_ctx`. Ollama uses its configured default instead of the computed window; a default of `1024` therefore reduces the window from 2048 to 1024.

Return the computed positive value:

```suggestion
            num_ctx: targetLength,
```

Update the test and comment in `never exceeds the model limit even for tiny inputs` that currently expect an unset value. Add parameterized tests for model limits `512`, `1024`, and `2048` in `execute`, `toolsExecute`, and `embed`, without an explicit `num_ctx`. Assert the actual `options.num_ctx` passed to `ollama.chat` / `ollama.embed`, not only the helper's return value. Keep the existing explicit-override tests passing.

**Review by pfrankov — CHANGES_REQUESTED** [review]
The earlier requested fixes remain accepted. I missed one boundary case in my previous review, so I have withdrawn the approval. Please address the inline P2 issue and add outgoing-request regression tests before merge.

Reviewed `b57f8f3` against `9ec77ec`; reproduced the difference with isolated sizing checks. These checks do not run the project test suite or a live Ollama server.

File: src/handlers/OllamaHandler.ts (hunk 4 above)
Comment (franzthiemann) [reply src/handlers/OllamaHandler.ts]: Fixed in a03f02b. Applied your suggestion: `optimizeContext()` now returns the clamped `targetLength` unconditionally.

You are right about the distinction that made this a real bug. An absent `num_ctx` is not a request for our default; it defers to Ollama's own configured default, which can be smaller than the window we just computed. Your 2048-limit case went out bare and a server default of 1024 would halve it.

Since `num_ctx` is now always a number, I also dropped the optional return from `optimizeContext()` and `applyContextOptimization()` and removed the `if (num_ctx)` guard in `resolveContextWindow()` — that guard had become unreachable, and leaving it would have quietly cost branch coverage.

Tests as requested: parameterized over limits 512, 1024 and 2048 in `execute`, `toolsExecute` and `embed`, asserting the `num_ctx` on the actual `ollama.chat` / `ollama.embed` call rather than the helper's return. All nine fail against the previous behaviour and pass now. The explicit-override tests still pass, and `never exceeds the model limit even for tiny inputs` plus its comment now expect the value to be sent.

**Review by pfrankov — APPROVED** [review]
Re-reviewed `a03f02b1170ff3757c4bb904b69d2b3d345e0ec5`, including the latest fix relative to `b57f8f3`, the new regression tests, and the earlier requested changes.

**No blocking findings remain. Ready to merge. This approval supersedes my previous changes-requested review.**

- The last P2 is fixed: `optimizeContext()` always returns the clamped window, and `resolveContextWindow()` sends it without the former omission. The nine new tests inspect the actual `ollama.chat` / `ollama.embed` arguments for limits 512, 1024 and 2048 across `execute`, `toolsExecute` and `embed`.
- Explicit `options.num_ctx` values remain authoritative in both chat paths. The existing override regressions are retained.
- The unsupported prompt-truncation check remains removed. Generation-limit reporting stays separate from the returned text and does not infer dropped input from token counts.
- The current three-commit history and PR description no longer contain AI-authorship credits.

Validation: source review plus 29 isolated Node checks using copied sizing/streaming method excerpts with mocked dependencies. These passed, but are not a run of the repository's Vitest suite, build, or a live Ollama server. The project-level results and the i18n/localStorage failure on clean master remain author-reported; I did not independently reproduce them. This HEAD has no GitHub check runs.

No further code changes requested.
